### PR TITLE
Update EIP-4972: Move status from draft to review

### PR DESCRIPTION
**Pull request by qizhou**
Change status from draft to review

When opening a pull request to submit a new EIP, please use the suggested template: https://github.com/ethereum/EIPs/blob/master/eip-template.md

We have a GitHub bot that automatically merges some PRs. It will merge yours immediately if certain criteria are met:

 - The PR edits only existing draft PRs.
 - The build passes.
 - Your GitHub username or email address is listed in the 'author' header of all affected PRs, inside <triangular brackets>.
 - If matching on email address, the email address is the one publicly listed on your GitHub profile.

### CODE REVIEW

**Review by Pandapip1 — COMMENTED** [review]
The following is a change I feel needs to be made:

1. For `IERC721NOA`, **specify what happens when someone calls `ownerOf` on a `tokenId` that is owned by an NOA.** Should it throw? Return the zero address? This is ambiguous right now.

Here are some suggestions for other changes that I feel the standard is lacking:

3. Add methods to `IERC721NOA` and `IERC20NOA` that allow users to transfer tokens *to* NOAs, and get the balance of NOAs.
4. Add an NOA extension for EIP-1155.

**Comment by qizhou** [comment]
> The following is a change I feel needs to be made:
> 
> 1. For `IERC721NOA`, **specify what happens when someone calls `ownerOf` on a `tokenId` that is owned by an NOA.** Should it throw? Return the zero address? This is ambiguous right now.
> 
> Here are some suggestions for other changes that I feel the standard is lacking:
> 
> 2. Add methods to `IERC721NOA` and `IERC20NOA` that allow users to transfer tokens _to_ NOAs, and get the balance of NOAs.
> 3. Add an NOA extension for EIP-1155. This is slightly less commonly used as EIP-721, but is common enough that it deserves to be included IMO.

Many thanks for the comments.  We are working on the comments and will update the PR shortly.

**Comment by SamWilsn** [comment]
Would you mind also rebasing on top of `master`? The GitHub workflow file in this branch is out of date (and is missing some checks.)

**Comment by qizhou** [comment]
> Would you mind also rebasing on top of `master`? The GitHub workflow file in this branch is out of date (and is missing some checks.)

Done.  Thanks for the suggestion.

**Comment by Pandapip1** [comment]
The eipw errors need to be fixed.

**Comment by qizhou** [comment]
> The following is a change I feel needs to be made:
> 
> 1. For `IERC721NOA`, **specify what happens when someone calls `ownerOf` on a `tokenId` that is owned by an NOA.** Should it throw? Return the zero address? This is ambiguous right now.

Thanks for the comment.  The  `ownerOf` method will return a name-derived address (NDA, also 20 bytes EH address).  The NDA can be obtained by calling `addressOfName` method.

> 
> Here are some suggestions for other changes that I feel the standard is lacking:
> 
> 2. Add methods to `IERC721NOA` and `IERC20NOA` that allow users to transfer tokens _to_ NOAs, and get the balance of NOAs.

Thanks for the comment.  Currently, it can be done by transferring tokens to NDAs (and getting the balances from the NDAs)

> 3. Add an NOA extension for EIP-1155. This is slightly less commonly used as EIP-721, but is common enough that it deserves to be included IMO.
Thanks for the comment. The extension for EIP-11155 is added.  Please take a look.

**Review by SamWilsn — COMMENTED** [review]
Couple other notes I can't directly comment on:
 - Since a NOA is a contract, technically its a subtype of a contract account (as opposed to alongside as specified in the abstract.)

File: EIPS/eip-4972.md [hunk 1]
```diff
@@ -57,7 +57,7 @@ interface INameOwnedAccount {
 }
 ```
 
-#### IERC721NOA
+#### IEIP-721NOA
```
[the remark this answers is not in the document]
Comment (qizhou) [reply EIPS/eip-4972.md]: Thanks for the comment.  Update the PR accordingly.

**Comment by qizhou** [comment]
> Couple other notes I can't directly comment on:
> 
> * Since a NOA is a contract, technically its a subtype of a contract account (as opposed to alongside as specified in the abstract.)

Thanks for the comment.  The NOA is not a contract, but an account, whose owner is managed by a name service. Further, we use `address(nameStringWithNSSuffix)` as the name-derived address so that we can lookup the operators/owners of the address.  Note that the address cannot be directly used by an external account or a contract in Ethereum - it can be only used by operators (as managed NS) via `transferFrom` type of transactions. Hope this can explain it.  Please let me know if you have further comments!

File: EIPS/eip-4972.md [hunk 2]
```diff
@@ -36,113 +36,112 @@ An NOA has
...
-where the username is the string representation of the username and nameService is the name service contract address. This will decouple the implementation from specific name services such as ENS.
 
-### Name Derived Address (INameOwnedAccount.addressOfName())
-With the bytes format mentioned above, we can follow the similar rule of CREATE2 opcode to compute the NOA address from nameService and hash of the username as `address(keccak256(0xff, keccak256(“eip-4972.addressOfName”), nameService, keccak256(username)))`.  This can ensure it won’t collide with existing smart contract account addresses.
+This can ensure it will not collide with existing smart contract account addresses.
 
 ### Ownership of a Name (INameOwnedAccount.isNameOwner())
-Normally we can get the owner from the name service and compare it with the message sender. We recommend the name service to define an owner function in the same format as ENS.
+Given an NDA, we need to lookup the owner from a name service and compare it with the message sender. Note that current ENS only supports namehash (32 bytes) to owner translation, and thus we may need an additional helper contract to map an NDA to a namehash for ENS.
```
Comment (Pandapip1) [inline EIPS/eip-4972.md:141]:
```suggestion

Given an NDA, we need to look up the owner from a name service and compare it with the message sender. Note that currently, ENS only supports namehash (32 bytes) to owner translation, and thus we may need an additional helper contract to map an NDA to a namehash for ENS.
```

File: EIPS/eip-4972.md [hunk 3]
```diff
@@ -36,113 +36,112 @@ An NOA has
...
 
 ## Security Considerations
 No security considerations were found.
```
Comment (Pandapip1) [inline EIPS/eip-4972.md:144]:
```suggestion

No security considerations were found.
```

File: EIPS/eip-4972.md [hunk 4]
```diff
@@ -36,113 +36,112 @@ An NOA has
...
+function allowance(address _owner, address _spender) public view returns (uint256 remaining)
+```
+
+## Rationale
+The isNameOwner function is sufficient for authenticating the message sender. One can verify the owner by looking up the owner of an NDA from a name service.
+
+The addressOfName interface decouples the implementation from specific NDA encoding algorithm. How to encode a name to an NDA is not defined by this standard, as long as the generated NDA does not collide with EOA/CA address space.
 
 ## Backwards Compatibility
 The new account type is compatible with existing ERC token standards.
```
Comment (Pandapip1) [inline EIPS/eip-4972.md:128]:
```suggestion

The new account type is compatible with existing ERC token standards.
```

File: EIPS/eip-4972.md [hunk 5]
```diff
@@ -36,113 +36,112 @@ An NOA has
...
 
 ## Reference Implementation
-### Name Format
-The decoded format of bytes name is not defined at this standard. One straightforward implementation would be:
+### Name Derived Address (INameOwnedAccount.addressOfName())
+Each token contract can decide how to encode the name to an NDA as long as the address space of NDAs does not overlap with existing EOA/CA spaces. We can follow the similar rule of CREATE2 opcode to compute the NDA from a name service and the hash of a username as
```
Comment (Pandapip1) [inline EIPS/eip-4972.md:132]:
```suggestion

Each token contract can decide how to encode the name to an NDA as long as the address space of NDAs does not overlap with existing EOA/CA spaces. We can follow the similar rule of CREATE2 opcode to compute the NDA from a name service and the hash of a username as 
```

File: EIPS/eip-4972.md [hunk 6]
```diff
@@ -36,113 +36,112 @@ An NOA has
...
 
 ## Reference Implementation
-### Name Format
-The decoded format of bytes name is not defined at this standard. One straightforward implementation would be:
+### Name Derived Address (INameOwnedAccount.addressOfName())
+Each token contract can decide how to encode the name to an NDA as long as the address space of NDAs does not overlap with existing EOA/CA spaces. We can follow the similar rule of CREATE2 opcode to compute the NDA from a name service and the hash of a username as 
+
 ```
-bytes memory name = abi.encode((string, ‘address’), (username, nameService))
+address(keccak256(0xff, keccak256(“eip-4972.addressOfName”), nameServiceAddress, keccak256(username))).
```
Comment (Pandapip1) [inline EIPS/eip-4972.md:135]:
```suggestion
address(keccak256(0xff, keccak256("eip-4972.addressOfName"), nameServiceAddress, keccak256(username))).
```

File: EIPS/eip-4972.md [hunk 7]
```diff
@@ -36,113 +36,112 @@ An NOA has
...
-    /// @dev Throws unless `msg.sender` is the owner of _owner. Throw if _owner is not
-    ///  the current owner. Emits the ApprovalForAll event. The contract MUST allow
-    ///  multiple operators per owner.
-    function setApprovalForAllFromName(bytes memory _owner, address _operator, bool _approved) public returns(bool);
-    
-    /// @notice This function returns true if interfaceId is the id of IERC721NOA
-    /// @dev see {IERC165-supportsInterface}
-    function supportsInterface(bytes4 interfaceId) external view returns(bool);
-}
+#### `IERC721NOA`
```
Comment (Pandapip1) [inline EIPS/eip-4972.md:62]:
```suggestion
#### `IERC721NOA`

```

File: EIPS/eip-4972.md [hunk 8]
```diff
@@ -36,113 +36,112 @@ An NOA has
...
-    ///  allowance with _value.
-    function approveFromName(bytes memory _owner, address _spender, uint256 _value) public returns(bool);
-    
-    /// @notice This function returns true if interfaceId is the id of IERC721NOA
-    /// @dev see {IERC165-supportsInterface}
-    function supportsInterface(bytes4 interfaceId) external view returns(bool);
-}
+The existing `transferFrom` and `safeBatchTransferFrom` functions MUST assume caller is approved to manage tokens of `_from` if `_from` represents a name and the caller is the owner of the name.
+
+```
```
Comment (Pandapip1) [inline EIPS/eip-4972.md:69]:
```suggestion
```solidity
```

File: EIPS/eip-4972.md [hunk 9]
```diff
@@ -36,113 +36,112 @@ An NOA has
...
-    ///  the current owner. Emits the ApprovalForAll event. The contract MUST allow
-    ///  multiple operators per owner.
-    function setApprovalForAllFromName(bytes memory _owner, address _operator, bool _approved) public returns(bool);
-    
-    /// @notice This function returns true if interfaceId is the id of IERC721NOA
-    /// @dev see {IERC165-supportsInterface}
-    function supportsInterface(bytes4 interfaceId) external view returns(bool);
-}
+#### `IERC721NOA`
+```
```
Comment (Pandapip1) [inline EIPS/eip-4972.md:63]:
```suggestion
```solidity
```

File: EIPS/eip-4972.md [hunk 10]
```diff
@@ -36,113 +36,112 @@ An NOA has
...
 
 
 ### Name Representation
-A name is represented by a bytes array which is ABI encoded. It **MAY** contain metadata of the name such as the name service the name belongs to.  Examples of the name are "vitalik.eth", "vitalik@0x00000000000C2E074eC69A0dFb2997BA6C7d2e1e", or "qizhou.fb".
+A name is represented by a bytes array which is ABI encoded. It **MAY** contain metadata of the name such as the name service the name belongs to.  Example of the name is "vitalik.eth" or "qizhou@quarkchain.org".
 
 ### Interface
 #### INameOwnedAccount
 ```
```
Comment (Pandapip1) [inline EIPS/eip-4972.md:47]:
```suggestion
```solidity
```

File: EIPS/eip-4972.md [hunk 11]
```diff
@@ -36,113 +36,112 @@ An NOA has
...
 ## Backwards Compatibility
 The new account type is compatible with existing ERC token standards.
 
 ## Reference Implementation
-### Name Format
-The decoded format of bytes name is not defined at this standard. One straightforward implementation would be:
+### Name Derived Address (INameOwnedAccount.addressOfName())
+Each token contract can decide how to encode the name to an NDA as long as the address space of NDAs does not overlap with existing EOA/CA spaces. We can follow the similar rule of CREATE2 opcode to compute the NDA from a name service and the hash of a username as 
+
 ```
```
Comment (Pandapip1) [inline EIPS/eip-4972.md:134]:
```suggestion
```solidity
```

File: EIPS/eip-4972.md [hunk 12]
```diff
@@ -36,113 +36,112 @@ An NOA has
...
+
+## Rationale
+The isNameOwner function is sufficient for authenticating the message sender. One can verify the owner by looking up the owner of an NDA from a name service.
+
+The addressOfName interface decouples the implementation from specific NDA encoding algorithm. How to encode a name to an NDA is not defined by this standard, as long as the generated NDA does not collide with EOA/CA address space.
 
 ## Backwards Compatibility
 The new account type is compatible with existing ERC token standards.
 
 ## Reference Implementation
```
Comment (Pandapip1) [inline EIPS/eip-4972.md:130]:
```suggestion
## Reference Implementation

```

File: EIPS/eip-4972.md [hunk 13]
```diff
@@ -36,113 +36,112 @@ An NOA has
...
-### Name Format
-The decoded format of bytes name is not defined at this standard. One straightforward implementation would be:
+### Name Derived Address (INameOwnedAccount.addressOfName())
```
Comment (Pandapip1) [inline EIPS/eip-4972.md:131]:
```suggestion
### Name Derived Address (INameOwnedAccount.addressOfName())

```

File: EIPS/eip-4972.md [hunk 14]
```diff
@@ -36,113 +36,112 @@ An NOA has
...
 
-The addressOfName interface decouples the implementation from specific hashing algorithms, as long as the generated address doesn’t collide with EOA/CA address space.
+```
+function safeTransferFrom(address _from, address _to, uint256 _id, uint256 _value, bytes calldata _data) external;
+function safeBatchTransferFrom(address _from, address _to, uint256[] calldata _ids, uint256[] calldata _values, bytes calldata _data) external;
+```
+
+The existing `isApprovedForAll` function of `IERC1155` MUST return true if the `_owner` represents a name and `_operator` is the owner of the name.
+
+```
```
Comment (Pandapip1) [inline EIPS/eip-4972.md:101]:
```suggestion
```solidity
```

File: EIPS/eip-4972.md [hunk 15]
```diff
@@ -36,113 +36,112 @@ An NOA has
...
+function isApprovedForAll(address _owner, address _operator) external view returns (bool);
+```
+
+#### `IERC20NOA`
```
Comment (Pandapip1) [inline EIPS/eip-4972.md:105]:
```suggestion
#### `IERC20NOA`

```

File: EIPS/eip-4972.md [hunk 16]
```diff
@@ -36,113 +36,112 @@ An NOA has
...
+```
```
Comment (Pandapip1) [inline EIPS/eip-4972.md:106]:
```suggestion
```solidity
```

File: EIPS/eip-4972.md [hunk 17]
```diff
@@ -36,113 +36,112 @@ An NOA has
...
+interface IERC20NOA is IERC20, INameOwnedAccount {}
+```
+
+The existing `transferFrom` function MUST assume caller is approved to manage tokens of `_from` if `_from` represents a name and the caller is the owner of the name.
+
+```
```
Comment (Pandapip1) [inline EIPS/eip-4972.md:112]:
```suggestion
```solidity
```

File: EIPS/eip-4972.md [hunk 18]
```diff
@@ -36,113 +36,112 @@ An NOA has
...
+function transferFrom(address _from, address _to, uint256 _value) public returns (bool success)
+```
+
+The existing `allowance ` function MUST return max value of uint256 if `_owner` represents a name and the `_spender` is the owner of the name.
+
+```
```
Comment (Pandapip1) [inline EIPS/eip-4972.md:118]:
```suggestion
```solidity
```

File: EIPS/eip-4972.md [hunk 19]
```diff
@@ -36,113 +36,112 @@ An NOA has
...
+function allowance(address _owner, address _spender) public view returns (uint256 remaining)
+```
+
+## Rationale
```
Comment (Pandapip1) [inline EIPS/eip-4972.md:122]:
```suggestion
## Rationale

```

File: EIPS/eip-4972.md [hunk 20]
```diff
@@ -36,113 +36,112 @@ An NOA has
...
+The name should be human-readable and can be easily recognized socially. An example is an username of a name service or an email address of a centralized platform such as Gmail. The name-derived address (NDA) is a normal Ethereum address that should not collide with the existing addresses of EOA/CA. Since we cannot use NDA as msg.sender, the right to transfer the tokens of the NDA is controlled by the owner/owners of the name. The name to owner/owners mapping is managed by an on-chain name service, and the owner/owners should be EOA/CA, which can be the addresses of 3-rd custodians (e.g., Facebook or Google) or self-custodian. By changing the owner of the name to the EOA of the user (can be done by requesting the custodian), the NDA becomes self-custodian, and no one should be able to transfer the assets unless the approved by the self-custodian user. 
 
 
 ### Name Representation
```
Comment (Pandapip1) [inline EIPS/eip-4972.md:42]:
```suggestion
### Name Representation

```

File: EIPS/eip-4972.md [hunk 21]
```diff
@@ -36,113 +36,112 @@ An NOA has
...
-A name is represented by a bytes array which is ABI encoded. It **MAY** contain metadata of the name such as the name service the name belongs to.  Examples of the name are "vitalik.eth", "vitalik@0x00000000000C2E074eC69A0dFb2997BA6C7d2e1e", or "qizhou.fb".
+A name is represented by a bytes array which is ABI encoded. It **MAY** contain metadata of the name such as the name service the name belongs to.  Example of the name is "vitalik.eth" or "qizhou@quarkchain.org".
 
 ### Interface
```
Comment (Pandapip1) [inline EIPS/eip-4972.md:45]:
```suggestion
### Interface

```

File: EIPS/eip-4972.md [hunk 22]
```diff
@@ -36,113 +36,112 @@ An NOA has
...
 #### INameOwnedAccount
```
Comment (Pandapip1) [inline EIPS/eip-4972.md:46]:
```suggestion
#### INameOwnedAccount

```

File: EIPS/eip-4972.md [hunk 23]
```diff
@@ -36,113 +36,112 @@ An NOA has
...
+```
+
+The existing `ownerOf` function at `IERC721` should return the NOA owning the token.
+
+```
+function ownerOf(uint256 _tokenId) external view returns (address);
+```
+
+#### `IERC1155NOA`
+```
```
Comment (Pandapip1) [inline EIPS/eip-4972.md:88]:
```suggestion
```solidity
```

File: EIPS/eip-4972.md [hunk 24]
```diff
@@ -36,113 +36,112 @@ An NOA has
...
+function safeTransferFrom(address _from, address _to, uint256 _tokenId) external payable;
 ```
 
-### Authentication
-The transfer and approve function is authenticated if and only if the message sender is the owner of the name.
+The existing `isApprovedForAll` function of `IERC721` MUST return true if the `_owner` represents a name and `_operator` is the owner of the name.
 
-## Rationale
-We use bytes array to represent a name to ensure it’s flexible enough to deal with different use cases. E.g. one can encode the name service contract address the name belongs to into the bytes array. One can also encode extra authentication data, such as zero knowledge proofs, into the bytes array. In the future, we may propose a standard to formalize the name for wider adoption.
+```
```
Comment (Pandapip1) [inline EIPS/eip-4972.md:77]:
```suggestion
```solidity
```

File: EIPS/eip-4972.md [hunk 25]
```diff
@@ -36,113 +36,112 @@ An NOA has
...
+function isApprovedForAll(address _owner, address _operator) external view returns (bool);
+```
+
+The existing `ownerOf` function at `IERC721` should return the NOA owning the token.
+
+```
```
Comment (Pandapip1) [inline EIPS/eip-4972.md:83]:
```suggestion
```solidity
```

File: EIPS/eip-4972.md [hunk 26]
```diff
@@ -36,113 +36,112 @@ An NOA has
...
+#### `IERC1155NOA`
+```
+interface IERC1155NOA is IERC1155, INameOwnedAccount  {}
+```
 
-The isNameOwner function is sufficient for authenticating the message sender. One can verify the owner by looking up the name owner from a name service, or check zero knowledge proofs encoded in name to prove the ownership directly without looking up anything.
+The existing `safeTransferFrom` and `safeBatchTransferFrom` functions MUST assume caller is approved to manage tokens of `_from` if `_from` represents a name and the caller is the owner of the name.
 
-The addressOfName interface decouples the implementation from specific hashing algorithms, as long as the generated address doesn’t collide with EOA/CA address space.
+```
```
Comment (Pandapip1) [inline EIPS/eip-4972.md:94]:
```suggestion
```solidity
```

**Review by Pandapip1 — CHANGES_REQUESTED** [review]

**Comment by qizhou** [comment]
Many thanks for the comments @Pandapip1 .  We are working on some issues found by the draft, and we will update a final one for review shortly.

**Review by Pandapip1 — CHANGES_REQUESTED** [review]
Please fix the merge conflicts

**Review by xinbenlv — APPROVED** [review]

**Comment by Pandapip1** [comment]
@qizhou please resolve the merge conflicts

**Comment by qizhou** [comment]
> @qizhou please resolve the merge conflicts

Thanks for rebasing with some latest changes.

**Comment by qizhou** [comment]
After discussing with the team, we will withdraw the version from "review" since there are a couple of major concerns.  We are focusing on polishing EIP-4972 and will create another PR once it is ready.  Many thanks for the comments and suggestions @Pandapip1 @SamWilsn @xinbenlv

**Pull request closed without merging**